src/include/migraphx/op/slice.hpp

```diff
@@ -98,37 +100,75 @@ struct slice
 
     shape normalize_compute_shape(std::vector<shape> inputs) const
     {
-        auto input_shape        = inputs[0];
-        auto t                  = input_shape.type();
-        const auto& old_lens    = input_shape.lens();
-        const auto& old_strides = input_shape.strides();
+        check_shapes{inputs, *this, true}.has(1);
+        auto input_shape = inputs[0];
+        auto t           = input_shape.type();
+        // For a static shape, old_lens will be adjusted to a new size
+        // for those axes that are sliced.
+        // For dynamic shape, the adjusted old_lens become the new max values,
+        // while retaining the old mins and opts if possible.
+        std::vector<std::size_t> new_mins;
+        std::vector<std::size_t> new_opts;
+        std::vector<std::size_t> old_lens;
+        std::vector<std::size_t> old_strides;
+        if(input_shape.dynamic())
+        {
+            old_lens = input_shape.max_lens();
+            new_mins = input_shape.min_lens();
+            new_opts = input_shape.opt_lens();
+        }
+        else
+        {
+            old_lens = input_shape.lens();
+            // For static shape (including eval after a dynamic input) the strides are indexed into
+            // the pre-slice array, so they are larger than the apparent size of the resulting
+            // shape.
+            old_strides = input_shape.strides();
+        }
 
         if(std::any_of(
-               axes.begin(), axes.end(), [&](auto i) { return (i >= old_lens.size() and i < 0); }))
+               axes.begin(), axes.end(), [&](auto i) { return (i >= old_lens.size() or i < 0); }))
         {
             MIGRAPHX_THROW("SLICE: input axis " + to_string_range(axes) + " out of range");
         }
 
         if(starts.size() != axes.size() or axes.size() != ends.size())
         {
-            MIGRAPHX_THROW("SLICE: inconsistent sizes");
+            MIGRAPHX_THROW("SLICE: inputs have inconsistent sizes");
         }
 
         std::vector<std::size_t> new_lens = old_lens;
         for(std::size_t i = 0; i < axes.size(); i++)
         {
             auto axis = axes[i];
-            new_lens[axis] =
+            auto sliced_length =
                 fix_index(old_lens, axis, ends[i]) - fix_index(old_lens, axis, starts[i]);
+            // A Numpy indexing convention: a slice size larger than the actual dimension
+            // is legal and is clipped to the axis size
+            new_lens[axis] = std::min(new_lens[axis], sliced_length);
```

Comment:
Doesn't look right.
Say `a` is a vector of length `n`.
In numpy you can do a slice `a[n-2:n+5]`. This would get you just the last value of `a`, so `length = 1`. The `sliced_length` calculated for this example would evaluate to `n+5 - n-2 = 3 -> length = 3`

Reply:
I think there's a bug with `fix_index()`. It's using `lens[axis]` rather than `lens[axis]-1` for the maximum index.

Reply:
Rechecked this, and the existing logic stands.  The maximum value in `fix_index()` is lens[axis] because it's really the max. "ends" value which is 1 more than the last index in the slice.  I will clarify in a comment.  In your example, the slice `a[n-2:n+5]` would result in `(a[n-2], a[n-1])` so 2 is the correct length.

Reply:
Adding a related fix: the existing code only calculates the maxes of the new output shape.  Mins need to be calculated separately too.